engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java

```diff
@@ -42,7 +42,7 @@ protected BuiltinRootNode(EnsoLanguage language) {
    * @return new node to use to call this builtin
    */
   public DirectCallNode createDirectCallNode() {
-    return DirectCallNode.create(getCallTarget());
+    return DirectCallNode.create(cloneUninitialized().getCallTarget());
```

Comment:
Cloning each `BuiltinRootNode` before use - builtins are lightweight and cloning helps to overcome Truffle recursive inlining limit.

Builtins that don't require `VirtualFrame` in their method signature are explicitly inlined by https://github.com/enso-org/enso/pull/6255#issuecomment-1506293252 - those that require `VirtualFrame` are "cloned" before each use. That makes Truffle believe that each call into the builtin is a call the different function.

Such a cloning is not as effective direct inlining, but certainly more effective than giving up on inlining completely due to recursive limit for inlining.